File: src/11.binarySearch/10.354.ts
export function maxEnvelopes(envelopes: number[][]): number {
    envelopes.sort((a, b) => {
        return a[0] === b[0] ? b[1] - a[1] : a[0] - b[0];
    });

    const nums = envelopes.map((item) => item[1]);
    const n = envelopes.length;
    const top = new Array(n).fill(0);
    let ans = 0;

    for (let i = 0; i < n; i++) {
        const num = nums[i];

        let l = 0;
        let r = ans;

        while (l < r) {
            const mid = l + Math.floor((r - l) / 2);
            if (top[mid] >= num) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }

        if (l == ans) {
            ans++;
        }

        top[l] = num;
    }

    return ans;
}

File: src/11.binarySearch/10.354.go
package array_greedy

import (
	"sort"
)

func MaxEnvelopes(envelopes [][]int) int {
	sort.Slice(envelopes, func(i, j int) bool {
		a, b := envelopes[i], envelopes[j]

		if a[0] == b[0] {
			return a[1] > b[1]
		}
		return a[0] < b[0]
	})

	// 开启套娃
	n := len(envelopes)
	heights := make([]int, n)
	for i := range heights {
		heights[i] = envelopes[i][1]
	}

	ans := 0
	top := make([]int, n)
	for _, num := range heights {
		l := 0
		r := ans

		for l < r {
			mid := l + (r-l)/2

			if top[mid] >= num {
				r = mid
			} else {
				l = mid + 1
			}
		}

		if l == ans {
			ans++
		}
		top[l] = num
	}

	return ans
}
